File: 1129. Shortest Path with Alternating Colors/Solution.py
class Solution:
    def shortestAlternatingPaths(self, n: int, redEdges: List[List[int]], blueEdges: List[List[int]]) -> List[int]:
        self.dist = 1
        self.res = [inf for i in range(n)]
        self.q = []
        redGraph, blueGraph = [[] for i in range(n)], [[] for i in range(n)]
        redVisited, blueVisited = [False for i in range(n)], [False for i in range(n)]
        self.res[0] = 0
        self.q.append([0, 0])
        self.q.append([0, 1])
        for x, y in redEdges: redGraph[x].append(y)
        for x, y in blueEdges: blueGraph[x].append(y)
        while self.q:
            l = len(self.q)
            for i in range(l):
                curr, col = self.q.pop(0)
                neighbours = blueGraph[curr] if col == 0 else redGraph[curr]
                for neighbour in neighbours:
                    if col == 0:
                        self.helper(blueVisited, neighbour, 1)
                    else:
                        self.helper(redVisited, neighbour, 0)
            self.dist += 1
        for i in range(n):
            if self.res[i] == inf: self.res[i] = -1
        return self.res

    def helper(self, visited, curr, nextCol):
        if visited[curr]: return
        visited[curr] = True
        self.res[curr] = min(self.res[curr], self.dist)
        self.q.append([curr, nextCol])

File: 1129. Shortest Path with Alternating Colors/Solution.java
class Solution {
    int dist;
    int[] res;
    List<Integer>[] redGraph, blueGraph;
    Queue<int[]> q;
    public int[] shortestAlternatingPaths(int n, int[][] redEdges, int[][] blueEdges) {
        dist = 1;
        res = new int[n];
        Arrays.fill(res, Integer.MAX_VALUE);
        redGraph = new List[n];
        blueGraph = new List[n];
        boolean[] redVisited = new boolean[n], blueVisited = new boolean[n];
        q = new LinkedList<>();
        q.offer(new int[] { 0, 0 });
        q.offer(new int[] { 0, 1 });
        res[0] = 0;
        redVisited[0] = true;
        blueVisited[0] = true;
        for (int i = 0; i < n; i++) {
            redGraph[i] = new ArrayList<>();
            blueGraph[i] = new ArrayList<>();
        }
        for (int[] r : redEdges) redGraph[r[0]].add(r[1]);
        for (int[] b : blueEdges) blueGraph[b[0]].add(b[1]);
        while (!q.isEmpty()) {
            int l = q.size();
            for (int i = 0; i < l; i++) {
                int[] curr = q.poll();
                List<Integer> neighbours = curr[1] == 0 ? blueGraph[curr[0]] : redGraph[curr[0]];
                for (int neighbour : neighbours) {
                    if (curr[1] == 0) {
                        helper(blueVisited, neighbour, 1);
                    } else {
                        helper(redVisited, neighbour, 0);
                    }
                }
            }
            dist++;
        }
        for (int i = 1; i < n; i++) {
            if (res[i] == Integer.MAX_VALUE) res[i] = -1;
        }
        return res;
    }
    public void helper(boolean[] visited, int curr, int nextColor) {
        if (visited[curr]) return;
        visited[curr] = true;
        res[curr] = Math.min(res[curr], dist);
        q.offer(new int[] { curr, nextColor });
    }
}